File: toggly-ruby/toggly/lib/toggly/definition_cache.rb
# frozen_string_literal: true

module Toggly
  # Shared definition-refresh cache helpers (hit/miss classification).
  module DefinitionCache
    module_function

    # Normalize an ETag/revision for comparison (strip weak prefix + quotes).
    #
    # @param etag [String, nil]
    # @return [String, nil]
    def normalize_etag(etag)
      return nil if etag.nil?

      trimmed = etag.to_s.strip
      return nil if trimmed.empty?

      trimmed = trimmed[2..].to_s.strip if trimmed.length >= 2 && %w[W w].include?(trimmed[0]) && trimmed[1] == "/"
      return trimmed[1...-1] if trimmed.length >= 2 && trimmed.start_with?('"') && trimmed.end_with?('"')

      trimmed
    end

    # @param left [String, nil]
    # @param right [String, nil]
    # @return [Boolean]
    def etags_match?(left, right)
      a = normalize_etag(left)
      b = normalize_etag(right)
      !a.nil? && !b.nil? && a == b
    end

    # @param left [String, nil]
    # @param right [String, nil]
    # @return [Boolean]
    def last_modified_match?(left, right)
      a = left&.to_s&.strip
      b = right&.to_s&.strip
      !a.nil? && !b.nil? && !a.empty? && a == b
    end

    # Match Go isCachedRevision: currentTS > 0 && incomingTS <= currentTS → hit.
    #
    # @param current_ts [Integer, nil]
    # @param incoming_ts [Integer, nil]
    # @return [Boolean]
    def cached_signed_timestamp?(current_ts, incoming_ts)
      return false if incoming_ts.nil?

      current = current_ts.to_i
      incoming = incoming_ts.to_i
      current.positive? && incoming <= current
    end

    # Classify an HTTP definitions response for cache telemetry.
    #
    # @param status_code [Integer]
    # @param existing_etag [String, nil]
    # @param response_etag [String, nil]
    # @param existing_last_modified [String, nil]
    # @param response_last_modified [String, nil]
    # @return [Symbol] :not_modified, :same_revision, :new_content, or :error_status
    def classify_http(
      status_code,
      existing_etag,
      response_etag,
      existing_last_modified: nil,
      response_last_modified: nil
    )
      return :not_modified if status_code == 304
      return :error_status if status_code != 200
      return :same_revision if etags_match?(existing_etag, response_etag)
      return :same_revision if last_modified_match?(existing_last_modified, response_last_modified)

      :new_content
    end
  end
end

File: toggly-rust/toggly/src/definition_cache.rs
//! Shared definition-refresh cache hit/miss helpers.

/// Outcome of one definition-refresh network attempt.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RefreshCacheOutcome {
    /// Served from local/cache without applying a new revision.
    Hit,
    /// Applied a new revision from the network.
    Miss,
}

/// HTTP classification for definitions refresh responses.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HttpCacheKind {
    /// HTTP 304 Not Modified.
    NotModified,
    /// HTTP 200 whose ETag or Last-Modified matches the existing revision.
    SameRevision,
    /// HTTP 200 with new content to apply.
    NewContent,
    /// Non-success status (caller treats as error / optional hit if cached).
    ErrorStatus,
}

/// Normalize an ETag/revision for comparison (strip weak prefix + quotes).
pub fn normalize_etag(etag: Option<&str>) -> Option<String> {
    let trimmed = etag?.trim();
    if trimmed.is_empty() {
        return None;
    }
    let mut value = trimmed;
    if value.len() >= 2
        && (value.as_bytes()[0] == b'W' || value.as_bytes()[0] == b'w')
        && value.as_bytes()[1] == b'/'
    {
        value = value[2..].trim();
    }
    if value.len() >= 2 && value.starts_with('"') && value.ends_with('"') {
        return Some(value[1..value.len() - 1].to_string());
    }
    Some(value.to_string())
}

/// True when both ETags/revisions normalize to the same non-empty value.
pub fn etags_match(left: Option<&str>, right: Option<&str>) -> bool {
    match (normalize_etag(left), normalize_etag(right)) {
        (Some(a), Some(b)) => a == b,
        _ => false,
    }
}

/// True when both Last-Modified values are present and equal.
pub fn last_modified_match(left: Option<&str>, right: Option<&str>) -> bool {
    match (left.map(str::trim), right.map(str::trim)) {
        (Some(a), Some(b)) if !a.is_empty() && !b.is_empty() => a == b,
        _ => false,
    }
}

/// Match Go/Ruby: `current_ts > 0 && incoming_ts <= current_ts` → cached hit.
pub fn cached_signed_timestamp(current_ts: Option<i64>, incoming_ts: i64) -> bool {
    match current_ts {
        Some(current) if current > 0 && incoming_ts <= current => true,
        _ => false,
    }
}

/// Classify an HTTP definitions response for cache telemetry.
pub fn classify_http(
    status_code: u16,
    existing_etag: Option<&str>,
    response_etag: Option<&str>,
    existing_last_modified: Option<&str>,
    response_last_modified: Option<&str>,
) -> HttpCacheKind {
    if status_code == 304 {
        return HttpCacheKind::NotModified;
    }
    if status_code != 200 {
        return HttpCacheKind::ErrorStatus;
    }
    if etags_match(existing_etag, response_etag)
        || last_modified_match(existing_last_modified, response_last_modified)
    {
        return HttpCacheKind::SameRevision;
    }
    HttpCacheKind::NewContent
}

/// Optional recorder for definition-refresh cache counters on the usage pipeline.
pub trait DefinitionCacheRecorder: Send + Sync {
    /// Count a refresh served from local cache / unchanged revision.
    fn record_definition_cache_hit(&self);
    /// Count a refresh that applied a new revision.
    fn record_definition_cache_miss(&self);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn matches_weak_and_strong_etags() {
        assert!(etags_match(Some("\"1\""), Some("W/\"1\"")));
        assert!(!etags_match(Some("\"1\""), Some("\"2\"")));
        assert!(!etags_match(None, Some("\"1\"")));
    }

    #[test]
    fn matches_last_modified_and_signed_timestamps() {
        let lm = "Mon, 01 Jan 2024 00:00:00 GMT";
        assert!(last_modified_match(Some(lm), Some(lm)));
        assert!(!last_modified_match(Some(lm), None));
        assert!(cached_signed_timestamp(Some(100), 100));
        assert!(cached_signed_timestamp(Some(100), 50));
        assert!(!cached_signed_timestamp(Some(100), 101));
        assert!(!cached_signed_timestamp(Some(0), 50));
        assert!(!cached_signed_timestamp(None, 50));
    }

    #[test]
    fn classifies_http_responses() {
        assert_eq!(
            classify_http(304, Some("\"1\""), None, None, None),
            HttpCacheKind::NotModified
        );
        assert_eq!(
            classify_http(200, Some("\"1\""), Some("\"1\""), None, None),
            HttpCacheKind::SameRevision
        );
        assert_eq!(
            classify_http(200, Some("\"1\""), Some("\"2\""), None, None),
            HttpCacheKind::NewContent
        );
        assert_eq!(
            classify_http(500, Some("\"1\""), None, None, None),
            HttpCacheKind::ErrorStatus
        );
        let lm = "Mon, 01 Jan 2024 00:00:00 GMT";
        assert_eq!(
            classify_http(200, Some("\"1\""), Some("\"2\""), Some(lm), Some(lm)),
            HttpCacheKind::SameRevision
        );
    }
}
